### PR TITLE
Schedulers & optimizers

### PR DESCRIPTION
**Pull request by juliakreutzer**
- Refactoring: building of schedulers, optimizers, gradient clippers now in builders.py
- Extension: more options for optimizers; updated in default config
- Bug fix: `self.scheduler.step()` is called either after validation or at the beginning of each epoch, addressing issue #18.

### CODE REVIEW

**Comment by juliakreutzer** [comment]
Further enhancements:
- Refactoring: Lots of it. Important: renaming `criterion` to `log_function` and removing it from the configs since we only have one loss implemented (cross-entropy). 
- Bug fix: issue #5 fixed by setting `minimize_metric` according to value of `early_stopping_metric` and `eval_metric`.
